File: course/section_2_two_pointers/code/lesson_20_palindrome_linked_list.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def isPalindrome2(self, head: Optional[ListNode]) -> bool:
        vals = []
        current = head
        while current:
            vals.append(current.val)
            current = current.next
        return vals == vals[::-1]

    def reverse(self, head):
        previous = None
        current = head

        while current:
            next_node = current.next
            current.next = previous
            previous = current
            current = next_node

        return previous

    def isPalindrome(self, head):

        slow = fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        left = head
        right = self.reverse(slow)

        while right:
            if left.val != right.val:
                return False

            left = left.next
            right = right.next

        return True

File: course/section_2_two_pointers/code/lesson_20_palindrome_linked_list.php
<?php

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution {
    private function reverse($head) {
        $current = $head;
        $previous = null;
        while ($current) {
            $nextNode = $current->next;
            $current->next = $previous;
            $previous = $current;
            $current = $nextNode;
        }
        return $previous;
    }

    /**
     * @param ListNode $head
     * @return Boolean
     */
    function isPalindrome($head) {
        $slow = $head;
        $fast = $head;
        while ($fast && $fast->next) {
            $slow = $slow->next;
            $fast = $fast->next->next;
        }
        $left = $head;
        $right = $this->reverse($slow);
        while ($left and $right) {
            if ($left->val != $right->val) {
                return false;
            }
            $left = $left->next;
            $right = $right->next;
        }
        return true;
    }
}
